execution_engine/src/core/engine_state/genesis.rs

```diff
@@ -413,6 +406,898 @@ impl Distribution<ExecConfig> for Standard {
+                    )?),
+                );
+                round_seigniorage_rate_uref
+            };
+
+        let total_supply_uref = self
+            .uref_address_generator
+            .borrow_mut()
+            .new_uref(AccessRights::READ_ADD_WRITE);
+
+        let mut total_amount = U512::zero();
+
+        let purses: Vec<GenesisPurse> = self.create_purses()?;
+
+        let named_keys = {
+            let mut named_keys = NamedKeys::new();
+            named_keys.insert(
+                ROUND_SEIGNIORAGE_RATE_KEY.to_string(),
+                round_seigniorage_rate_uref.into(),
+            );
+
+            // associate purse urefs and balance urefs
+            for purse in purses.iter() {
+                match purse {
+                    GenesisPurse::ProofOfStake {
+                        purse_uref,
+                        balance_uref,
+                        amount,
+                    }
+                    | GenesisPurse::DelegatorReward {
+                        purse_uref,
+                        balance_uref,
+                        amount,
+                    }
+                    | GenesisPurse::ValidatorReward {
+                        purse_uref,
+                        balance_uref,
+                        amount,
+                    }
+                    | GenesisPurse::GenesisAccount {
+                        purse_uref,
+                        balance_uref,
+                        amount,
+                        ..
+                    }
+                    | GenesisPurse::GenesisValidator {
+                        purse_uref,
+                        balance_uref,
+                        amount,
+                        ..
+                    } => {
+                        total_amount = total_amount + amount;
+                        let purse_uref_name =
+                            purse_uref.remove_access_rights().to_formatted_string();
+                        let named_key = Key::URef(*balance_uref);
+                        named_keys.insert(purse_uref_name, named_key);
+                    }
+                }
+            }
+
+            named_keys.insert(TOTAL_SUPPLY_KEY.to_string(), total_supply_uref.into());
+
+            named_keys
+        };
+
+        self.tracking_copy.borrow_mut().write(
+            total_supply_uref.into(),
+            StoredValue::CLValue(
+                CLValue::from_t(total_amount)
+                    .map_err(|_| GenesisError::CLValue(TOTAL_SUPPLY_KEY.to_string()))?,
+            ),
+        );
+
+        let entry_points = self.mint_entry_points();
+
+        let access_key = self
+            .uref_address_generator
+            .borrow_mut()
+            .new_uref(AccessRights::READ_ADD_WRITE);
+
+        let (_, contract_hash) = self.store_contract(access_key, named_keys, entry_points);
+
+        Ok((contract_hash, purses))
+    }
+
+    pub fn create_proof_of_stake(
+        &self,
+        genesis_purses: &[GenesisPurse],
```

Comment:
We would likely use a different representation here, since we should be able to get rid of `GenesisPurse` with the aforementioned changes.

(and below)